src/main/java/com/fasterxml/jackson/core/JsonFactory.java

```diff
@@ -2181,12 +2181,24 @@ public RecyclerPool<BufferRecycler> _getRecyclerPool() {
      * @return I/O context created
      */
     protected IOContext _createContext(ContentReference contentRef, boolean resourceManaged) {
-        // 21-Mar-2021, tatu: Bit of defensive coding for backwards compatibility
+        BufferRecycler br;
+
         if (contentRef == null) {
             contentRef = ContentReference.unknown();
+            br = null;
+        } else {
+            Object content = contentRef.getRawContent();
+            // 18-Jan-2024, tatu: [core#1195] Let's see if we can reuse already allocated recycler
+            //   (is the case when SegmentedStringWriter / ByteArrayBuilder passed)
+            br = (content instanceof BufferRecycler.Gettable)
+                    ? ((BufferRecycler.Gettable) content).bufferRecycler()
+                    : null;
+        }
+        if (br == null) {
+            br = _getBufferRecycler();
         }
         return new IOContext(_streamReadConstraints, _streamWriteConstraints, _errorReportConfiguration,
-                _getBufferRecycler(), contentRef, resourceManaged);
+                br, contentRef, resourceManaged);
```

Comment:
This is very similar to what I did in my pull request, but in my opinion there is one important missing thing here: if the `BufferRecycler` hasn't been taken from the pool here but you're reusing the one coming from the `BufferRecycler.Gettable` [the IOContext should be aware that it is not the main owner of the resource](https://github.com/FasterXML/jackson-core/pull/1192/files#diff-6cb3e29fa26f4258ee5ad42c9a0a2bd91e053d2a7e7dbe16b67230dcf542ab14R2193) and [in this case it should NOT be its reposibility to put the `BufferRecycler` back into the pool when closed](https://github.com/FasterXML/jackson-core/pull/1192/files#diff-a252aa8da98b5f0fe4dfd326d160337c010be5921607f0a64df3bb9634779736R469).

Reply:
That's part that is not ideal: in this case ownership is assumed to be passed over. I'll have to think about this a bit.
For existing implementations difference does not matter but other pools could assume this important.

The reason it should not matter a lot, at this point, is due to synchronization in`BufferRecycler`.

Reply:
And yes, this is very similar: I liked many parts so wanted to start with those. And will likely copy the ownership part wrt `IOContext` too.

Reply:
Added possibility of external life-cycle.